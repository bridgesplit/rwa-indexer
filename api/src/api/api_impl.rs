use std::str::FromStr;

use open_rpc_derive::document_rpc;
use open_rpc_schema::document::OpenrpcDocument;
use rwa_types::rapi::{get_all_accounts, FullAccount};
use sea_orm::{ConnectionTrait, DbBackend, Statement};
use solana_sdk::pubkey::Pubkey;
use sqlx::postgres::PgPoolOptions;
use {
    crate::api::*,
    crate::config::Config,
    crate::error::RwaApiError,
    async_trait::async_trait,
    sea_orm::{DatabaseConnection, SqlxPostgresConnector},
};

pub struct RwaApi {
    db_connection: DatabaseConnection,
}

pub fn validate_pubkey(str_pubkey: String) -> Result<Pubkey, RwaApiError> {
    Pubkey::from_str(&str_pubkey).map_err(|_| RwaApiError::PubkeyValidationError(str_pubkey))
}

impl RwaApi {
    pub async fn from_config(config: Config) -> Result<Self, RwaApiError> {
        let pool = PgPoolOptions::new()
            .max_connections(250)
            .connect(&config.database_url)
            .await?;

        let conn = SqlxPostgresConnector::from_sqlx_postgres_pool(pool);
        Ok(RwaApi {
            db_connection: conn,
        })
    }
}

#[document_rpc]
#[async_trait]
impl ApiContract for RwaApi {
    // Liveness probe determines if the pod is healthy. Kubernetes will restart the pod if this fails.
    async fn liveness(self: &RwaApi) -> Result<(), RwaApiError> {
        Ok(())
    }

    // Readiness probe determines if the pod has capacity to accept traffic. Kubernetes will not route traffic to this pod if this fails.
    // We are essentially checking if there are DB connections available.
    async fn readiness(self: &RwaApi) -> Result<(), RwaApiError> {
        self.db_connection
            .execute(Statement::from_string(
                DbBackend::Postgres,
                "SELECT 1".to_string(),
            ))
            .await?;
        Ok(())
    }

    async fn get_all_accounts(
        self: &RwaApi,
        payload: GetAllAccounts,
    ) -> Result<FullAccount, RwaApiError> {
        let GetAllAccounts { id } = payload;
        let id_bytes = validate_pubkey(id.clone())?.to_bytes().to_vec();

        get_all_accounts(&self.db_connection, id_bytes)
            .await
            .map_err(Into::into)
    }
}
