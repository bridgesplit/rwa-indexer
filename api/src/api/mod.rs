use crate::error::RwaApiError;
use async_trait::async_trait;
use open_rpc_derive::{document_rpc, rpc};
use open_rpc_schema::schemars::JsonSchema;
use serde::{Deserialize, Serialize};

mod api_impl;
pub use api_impl::*;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, Default)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct GetAllAccounts {
    pub id: String,
}

#[document_rpc]
#[async_trait]
pub trait ApiContract: Send + Sync + 'static {
    async fn liveness(&self) -> Result<(), RwaApiError>;
    async fn readiness(&self) -> Result<(), RwaApiError>;

    #[rpc(
        name = "getAllAccounts",
        params = "named",
        summary = "Get all RWA accounts by its mint"
    )]
    async fn get_all_accounts(
        &self,
        payload: GetAllAccounts,
    ) -> Result<rwa_types::rapi::FullAccount, RwaApiError>;
}
