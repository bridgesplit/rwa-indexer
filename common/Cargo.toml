[package]
edition = "2021"
name = "common"
version = "0.1.0"

[dependencies]
figment = {workspace = true}
git2 = {workspace = true}
serde = {workspace = true}
sqlx =  {workspace = true}
