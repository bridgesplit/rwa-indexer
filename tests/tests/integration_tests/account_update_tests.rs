use function_name::named;

use itertools::Itertools;
use rwa_api::api::{self, ApiContract};

use serial_test::serial;

use super::common::*;

#[tokio::test]
#[serial]
#[named]
async fn test_get_rwa_accounts_by_mint() {
    let name = trim_test_name(function_name!());
    let setup = TestSetup::new(name.clone()).await;

    let seeds: Vec<SeedEvent> = vec![seed_token_mint(
        "Ea1yrC1xRXd6tWcHL4yhGRB31j6jTwdeqd3e9LHaYUwj",
    )];
    apply_migrations_and_delete_data(setup.db.clone()).await;
    index_seed_events(&setup, seeds.iter().collect_vec()).await;
    let request: api::GetRwaAccountsByMint = serde_json::from_str(
        r#"{
        "id": "Ea1yrC1xRXd6tWcHL4yhGRB31j6jTwdeqd3e9LHaYUwj"
    }"#,
    )
    .unwrap();
    let response = setup
        .rwa_api
        .get_rwa_accounts_by_mint(request)
        .await
        .unwrap();
    insta::assert_json_snapshot!(setup.name.clone(), response);
}

#[tokio::test]
#[serial]
#[named]
async fn test_get_rwa_accounts_by_authority() {
    let name = trim_test_name(function_name!());
    let setup = TestSetup::new(name.clone()).await;

    let seeds: Vec<SeedEvent> = vec![seed_token_mint(
        "dqci6TGXWGTki8DTQ1QcJ3qg4fPbxUzj6MwJBGjuhHX",
    )];
    apply_migrations_and_delete_data(setup.db.clone()).await;
    index_seed_events(&setup, seeds.iter().collect_vec()).await;
    let request: api::GetRwaAccountsByMint = serde_json::from_str(
        r#"{
        "id": "dqci6TGXWGTki8DTQ1QcJ3qg4fPbxUzj6MwJBGjuhHX"
    }"#,
    )
    .unwrap();
    let response = setup
        .rwa_api
        .get_rwa_accounts_by_authority(request)
        .await
        .unwrap();
    insta::assert_json_snapshot!(setup.name.clone(), response);
}

#[tokio::test]
#[serial]
#[named]
async fn test_get_rwa_accounts_by_delegate() {
    let name = trim_test_name(function_name!());
    let setup = TestSetup::new(name.clone()).await;

    let seeds: Vec<SeedEvent> = vec![seed_token_mint(
        "8tAP4FDpFehdQVmwTqCoYtDehG863TJqnTKZYz6UPR6m",
    )];
    apply_migrations_and_delete_data(setup.db.clone()).await;
    index_seed_events(&setup, seeds.iter().collect_vec()).await;
    let request: api::GetRwaAccountsByMint = serde_json::from_str(
        r#"{
        "id": "8tAP4FDpFehdQVmwTqCoYtDehG863TJqnTKZYz6UPR6m"
    }"#,
    )
    .unwrap();
    let response = setup
        .rwa_api
        .get_rwa_accounts_by_delegate(request)
        .await
        .unwrap();
    insta::assert_json_snapshot!(setup.name.clone(), response);
}

// get_rwa_accounts_by_delegate
