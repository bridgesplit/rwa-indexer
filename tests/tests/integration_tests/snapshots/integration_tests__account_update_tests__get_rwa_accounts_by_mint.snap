---
source: tests/tests/integration_tests/account_update_tests.rs
assertion_line: 33
expression: response
---
{
  "asset_controller": {
    "address": "8wpRqZ5993g1ubygtgUgh3bw8VEfp4LV7VuofL51pa4S",
    "mint": "Ea1yrC1xRXd6tWcHL4yhGRB31j6jTwdeqd3e9LHaYUwj",
    "authority": "dqci6TGXWGTki8DTQ1QcJ3qg4fPbxUzj6MwJBGjuhHX",
    "delegate": "8wpRqZ5993g1ubygtgUgh3bw8VEfp4LV7VuofL51pa4S",
    "version": 1,
    "closed": false
  },
  "data_registry": {
    "address": "ApfwesFjbtvR3qTXmRM8hGyQ9HiufRPsQQt5wUH9e8Uw",
    "mint": "Ea1yrC1xRXd6tWcHL4yhGRB31j6jTwdeqd3e9LHaYUwj",
    "version": 1,
    "closed": false
  },
  "identity_registry": {
    "address": "8tAP4FDpFehdQVmwTqCoYtDehG863TJqnTKZYz6UPR6m",
    "mint": "Ea1yrC1xRXd6tWcHL4yhGRB31j6jTwdeqd3e9LHaYUwj",
    "authority": "dqci6TGXWGTki8DTQ1QcJ3qg4fPbxUzj6MwJBGjuhHX",
    "delegate": "8tAP4FDpFehdQVmwTqCoYtDehG863TJqnTKZYz6UPR6m",
    "version": 1,
    "closed": false
  },
  "policy_engine": {
    "address": "EVfbhfNeH9TqcBVbr9TEGpQvKgR51vJSLiwZXvLoR7DG",
    "mint": "Ea1yrC1xRXd6tWcHL4yhGRB31j6jTwdeqd3e9LHaYUwj",
    "authority": "dqci6TGXWGTki8DTQ1QcJ3qg4fPbxUzj6MwJBGjuhHX",
    "delegate": "EVfbhfNeH9TqcBVbr9TEGpQvKgR51vJSLiwZXvLoR7DG",
    "policies": [
      "9ouBMZVbQeVLjpi6oewvM4aEc16cuQCB1mmwgrkVjCbL",
      "DMX8iMbWnw82gSDjMQ9fVYNNYFj4f3ceKd7vHi3x3eMG",
      "Cg5wADuxpSVQdAxhMwdCzDQzPwVLa6aTu3uG3eNTD2CY",
      "5ynb2SyJSxF17DRq8vvDSQVV1xKx6g8B5LsnzUwEnKRe",
      "11111111111111111111111111111111",
      "11111111111111111111111111111111",
      "11111111111111111111111111111111",
      "11111111111111111111111111111111",
      "11111111111111111111111111111111",
      "11111111111111111111111111111111"
    ],
    "version": 1,
    "closed": false
  }
}
